File: src/main/java/com/leetcode/graphs/NoOfIslands.java
package com.leetcode.graphs;

public class NoOfIslands {
    private void dfs(char[][] grid,int  row,int  col){

        if(grid[row][col]!='1'){
            return;
        }
        grid[row][col]='2';
        if(row+1<grid.length && grid[row+1][col]=='1'){
            dfs(grid, row+1, col);
        }
        if(row-1>=0 && grid[row-1][col]=='1'){
            dfs(grid, row-1, col);
        }
        if(col+1<grid[0].length && grid[row][col+1]=='1'){
            dfs(grid,row, col+1);
        }
        if(col-1>=0 && grid[row][col-1]=='1'){
            dfs(grid, row,col-1);
        }

    }
    public int numIslands(char[][] grid) {
        int res=0;
        for(int row=0; row<grid.length;row++){
            for(int col=0; col<grid[0].length;col++){
                if (grid[row][col]=='1'){
                    this.dfs(grid, row,col);
                    res++;
                }
            }
        }
        return res;
    }
}

File: python/noOfIslands.py
from typing import List

class Solution:
    def __init__(self):
        self.neighbours = [(0,1), (1,0), (0,-1), (-1,0)]
        self.rows = 0
        self.cols=0
    def dfs(self, grid, row, col):
        if row<0 or col<0 or row>=self.rows or col >=self.cols or grid[row][col]!='1':
            return 
        else:
            grid[row][col]='2'
            for r, c in self.neighbours:
                self.dfs(grid, row+r, col+c)

    def numIslands(self, grid: List[List[str]]) -> int:
        self.rows = len(grid)
        self.cols = len(grid[0])
        res=0
        for row in range(self.rows):
            for col in range(self.cols):
                if grid[row][col]=='1':
                    self.dfs(grid, row, col)
                    res+=1
        return res